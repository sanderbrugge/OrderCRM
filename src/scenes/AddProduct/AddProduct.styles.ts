import { StyleSheet } from "react-native";

const AddProductStyles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 10
  },
  header: {
    fontSize: 18,
    fontWeight: "600"
  },
  bodyContainer: {
    margin: 10
  }
});

export default AddProductStyles;
