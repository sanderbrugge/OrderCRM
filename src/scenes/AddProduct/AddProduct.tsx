import * as React from "react";
import { View, Text, ActivityIndicator, FlatList } from "react-native";
import { connect } from "react-redux";
import { Store } from "../../ducks";
import {
  actions as productActions,
  AsyncProducts
} from "../../ducks/product.reducer";
import { PENDING, SUCCESS, FAILURE, UNSENT } from "../../ducks/redux.types";
import { colors } from "../../styles/base";
import Error from "../../components/Error";
import ProductItem from "../../components/Product";
import { actions as orderActions } from "../../ducks/order.reducer";
import { Item } from "../../api/Order/order";

interface IProps {
  products: AsyncProducts;
  orderId: string;
  addProduct: (orderId: string, item: Item) => void;
  fetchProducts: () => void;
}

const AddProduct: React.FC<IProps> = ({
  products,
  fetchProducts,
  orderId,
  addProduct
}) => {
  React.useEffect(() => {
    if (products.status === UNSENT) {
      fetchProducts();
    }
  }, []);

  const isFetching = products.status === PENDING;
  const hasData = products.status === SUCCESS && products.data;
  const error = products.status === FAILURE;
  return (
    <View style={{ flex: 1, margin: 10 }}>
      <Text style={{ fontSize: 18, fontWeight: "600" }}>Add Product</Text>
      <View style={{ margin: 10 }}>
        {isFetching && <ActivityIndicator size="large" color={colors.grey} />}

        {hasData && (
          <FlatList
            data={products.data}
            renderItem={({ item }) => (
              <ProductItem
                product={item}
                orderId={orderId}
                orderProduct={addProduct}
              />
            )}
            keyExtractor={order => order.id}
          />
        )}

        {error && <Error message={products.error!.message} />}
      </View>
    </View>
  );
};

const mapStateToProps = (state: Store) => ({
  products: state.products
});

const mapDispatchToProps = (dispatch: any) => ({
  fetchProducts: () => dispatch(productActions.fetchProducts()),
  addProduct: (orderId: string, item: Item) =>
    dispatch(orderActions.addProduct(orderId, item))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddProduct);
