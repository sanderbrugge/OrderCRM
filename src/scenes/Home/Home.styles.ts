import { StyleSheet } from "react-native";
import { colors } from "../../styles/base";

const HomeStyles = StyleSheet.create({
  container: {
    height: 1000,
    flex: 1,
    backgroundColor: colors.white
  },
  row: {
    padding: 10,
    fontSize: 18,
    height: 44
  }
});

export default HomeStyles;
