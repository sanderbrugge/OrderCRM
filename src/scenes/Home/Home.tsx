import * as React from "react";
import { FlatList, View, ActivityIndicator } from "react-native";
import HomeStyles from "./Home.styles";
import Header from "../../components/Header";
import { connect } from "react-redux";
import { Store } from "../../ducks";
import {
  actions as orderActions,
  AsyncOrders,
  PENDING,
  SUCCESS,
  FAILURE
} from "../../ducks/order.reducer";
import { colors } from "../../styles/base";
import Error from "../../components/Error";
import OrderRowContainer from "../../components/OrderRow";

interface IHomeProps {
  orders: AsyncOrders;
  fetchOrders: () => void;
}

const Home: React.FC<IHomeProps> = ({ orders, fetchOrders }) => {
  React.useEffect(() => {
    fetchOrders();
  }, []);

  const isFetching = orders.status === PENDING;
  const hasData = orders.status === SUCCESS && orders.data;
  const error = orders.status === FAILURE;

  return (
    <View style={HomeStyles.container}>
      {isFetching && <ActivityIndicator size="large" color={colors.grey} />}

      {hasData && (
        <FlatList
          data={orders.data}
          renderItem={({ item }) => <OrderRowContainer order={item} />}
          keyExtractor={order => order.id}
        />
      )}

      {error && <Error message={orders.error!.message} />}
    </View>
  );
};

interface IProps {
  orders: AsyncOrders;
  fetchOrders: () => void;
}

const HomeWrapper: React.FC<IProps> = ({ orders, fetchOrders }) => (
  <Header
    title="All Orders"
    canNavigateBack={false}
    childView={<Home orders={orders} fetchOrders={fetchOrders} />}
  />
);

const mapStateToProps = (state: Store) => ({
  orders: state.orders
});

const mapDispatchToProps = (dispatch: any) => ({
  fetchOrders: () => dispatch(orderActions.fetchOrders())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomeWrapper);
