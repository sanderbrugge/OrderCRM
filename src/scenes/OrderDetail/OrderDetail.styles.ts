import { StyleSheet } from "react-native";
import { colors } from "../../styles/base";

const OrderDetailStyles = StyleSheet.create({
  itemContainer: {
    alignItems: "center",
    justifyContent: "center",
    alignSelf: "center",
    width: 200,
    height: 28,
    margin: 10,
    borderRadius: 8,
    backgroundColor: colors.blue
  },
  total: {
    color: colors.white,
    fontSize: 16
  },
  orderButtonContainer: {
    position: "absolute",
    bottom: 15,
    left: 20,
    right: 20
  }
});

export default OrderDetailStyles;
