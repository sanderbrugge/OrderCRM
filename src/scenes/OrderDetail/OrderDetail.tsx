import * as React from "react";
import { Order } from "../../api/Order/order";
import { Text, View } from "react-native";
import { NavigationInjectedProps, NavigationParams } from "react-navigation";
import Header from "../../components/Header";
import OrderButton from "../../components/OrderButton";
import { Icons } from "react-native-fontawesome";
import { colors } from "../../styles/base";
import ItemContainer from "../../components/Item";
import OrderDetailStyles from "./OrderDetail.styles";

interface IProps extends NavigationInjectedProps<NavigationParams> {}

const OrderDetail: React.FC<IProps> = ({ navigation }) => {
  const order = navigation.getParam("order", {}) as Order;
  return (
    <>
      <Header
        title="Order Overview"
        canNavigateBack
        childView={
          <>
            <View style={OrderDetailStyles.itemContainer}>
              <Text style={OrderDetailStyles.total}>TOTAL: ${order.total}</Text>
            </View>
            {order.items.map(item => (
              <ItemContainer key={item["product-id"]} item={item} />
            ))}
          </>
        }
      />
      <View style={OrderDetailStyles.orderButtonContainer}>
        <OrderButton
          title="ORDER"
          icon={Icons.shoppingCart}
          color={colors.blue}
          onClick={() => console.log("order placed")}
        />
      </View>
    </>
  );
};

export default OrderDetail;
