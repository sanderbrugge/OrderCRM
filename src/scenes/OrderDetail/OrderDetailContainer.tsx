import * as React from "react";
import { View } from "react-native";
import { NavigationInjectedProps, NavigationParams } from "react-navigation";
import Header from "../../components/Header";
import OrderButton from "../../components/OrderButton";
import { Icons } from "react-native-fontawesome";
import { colors } from "../../styles/base";
import OrderDetailStyles from "./OrderDetail.styles";
import HeaderBarButton from "../../components/HeaderBarButton/HeaderBarButton";
import RBSheet from "react-native-raw-bottom-sheet";
import { connect } from "react-redux";
import { Store } from "../../ducks";
import { AsyncOrders } from "../../ducks/order.reducer";
import { Order } from "../../api/Order/order";
import OrderDetail from "./OrderDetail";
import { placeOrder } from "../../api/Order/order.services";
import { actions as orderActions } from "../../ducks/order.reducer";

interface IProps extends NavigationInjectedProps<NavigationParams> {
  orders: AsyncOrders;
  removeProduct: (orderId: string, productId: string) => void;
}

/**
 * No need to use the API for this as all of the orders are available in Redux.
 *
 * @param orders all the orders available in the global state
 * @param id the ID to lookup
 */
function getOrderById(orders: Order[], id: string) {
  return orders.find(order => order.id === id);
}

/**
 * Also exported as named export to be able to test this component without the redux connection.
 *
 * @param param0 All orders, to select by ID and display it's items and the navigation object.
 */
export const OrderDetailContainer: React.FC<IProps> = ({
  orders,
  removeProduct,
  navigation
}) => {
  const order = React.useMemo(
    () => getOrderById(orders.data, navigation.getParam("id")),
    [orders, navigation]
  );
  const [isOrdering, setIsOrdering] = React.useState<boolean>(false);
  const [color, setColor] = React.useState<string>(colors.blue);
  const [icon, setIcon] = React.useState<string>(Icons.shoppingCart);
  const [title, setTitle] = React.useState<string>("ORDER");
  const disabled = isOrdering || (order && order.items.length === 0);

  const submitOrder = async () => {
    setIsOrdering(true);
    try {
      await placeOrder(order!.id);
      setColor(colors.green);
      setIcon(Icons.check);
      setTitle("PLACED");
    } catch (error) {
      console.log(error);
      setColor(colors.red);
      setIcon(Icons.times);
      setTitle("ERROR");
    } finally {
      setIsOrdering(false);
    }
  };

  const onPlaceOrder = React.useCallback(() => submitOrder(), [order]);

  const rbSheet = React.useRef<RBSheet | null>(null);

  return (
    <>
      <Header
        title="Order Overview"
        canNavigateBack
        rightBarButton={
          <HeaderBarButton
            icon={Icons.plus}
            onPress={() => rbSheet.current && rbSheet.current.open()}
          />
        }
        childView={
          <OrderDetail
            order={order!}
            rbSheet={rbSheet}
            removeProduct={removeProduct}
          />
        }
      />
      <View style={OrderDetailStyles.orderButtonContainer}>
        <OrderButton
          disabled={disabled}
          title={title}
          icon={icon}
          color={color}
          onClick={onPlaceOrder}
        />
      </View>
    </>
  );
};

const mapStateToProps = (state: Store) => ({
  orders: state.orders
});

const mapDispatchToProps = (dispatch: any) => ({
  removeProduct: (orderId: string, productId: string) =>
    dispatch(orderActions.removeProduct(orderId, productId))
});
/**
 * I could've added al the orders in the react navigation param as a dependency,
 * which would exclude this component to connect to Redux.
 *
 * But I opted against it because that would cause state management on different places, which is not optimal.
 */
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(OrderDetailContainer);
