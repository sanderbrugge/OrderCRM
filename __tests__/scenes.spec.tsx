import "react-native";
import * as React from "react";
import { OrderDetailContainer } from "../src/scenes/OrderDetail/OrderDetailContainer";
import { Home } from "../src/scenes/Home/Home";
import { AddProduct } from "../src/scenes/AddProduct/AddProduct";
import renderer from "react-test-renderer";
import { SUCCESS } from "../src/ducks/redux.types";

const orders = {
  status: SUCCESS,
  data: [
    {
      id: "2",
      "customer-id": "2",
      items: [
        {
          "product-id": "B102",
          quantity: "5",
          "unit-price": "4.99",
          total: "24.95"
        }
      ],
      total: "24.95"
    }
  ]
};

test("AddProduct renders correctly", async () => {
  const asyncProducts = {
    status: "UNSENT",
    data: []
  };

  const orderId = "1";
  const addProduct = jest.fn();
  const fetchProducts = jest.fn();

  const mockProps = {
    products: asyncProducts,
    orderId,
    addProduct,
    fetchProducts
  };

  // @ts-ignore
  const tree = await renderer.create(<AddProduct {...mockProps} />).toJSON();

  expect(tree).toMatchSnapshot();
});

jest.mock(
  "../src/components/OrderRow/OrderRowContainer",
  () => "orderRowContainer"
);

test("Home Component with orders renders correctly", async () => {
  const mockProps = {
    orders,
    fetchOrders: jest.fn()
  };

  // @ts-ignore
  const tree = await renderer.create(<Home {...mockProps} />).toJSON();

  expect(tree).toMatchSnapshot();
  expect(mockProps.fetchOrders).toBeCalledTimes(0);
});

// Mock this child component because it relies on react-navigation but can't find it.
jest.mock("../src/components/Header/Header", () => "header");

test("OrderDetail Component renders correctly", async () => {
  jest.mock("react-navigation", () => ({
    withNavigation: (component: React.FC) => component
  }));

  const mockProps = {
    navigation: { navigation: jest.fn(), getParam: jest.fn(), pop: jest.fn() },
    orders
  };

  const tree = await renderer
    // @ts-ignore
    .create(<OrderDetailContainer {...mockProps} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
