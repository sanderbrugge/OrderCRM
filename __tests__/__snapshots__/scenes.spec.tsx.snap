// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "margin": 10,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 18,
        "fontWeight": "600",
      }
    }
  >
    Add Product
  </Text>
  <View
    style={
      Object {
        "margin": 10,
      }
    }
  />
</View>
`;

exports[`Home Component with orders renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "customer-id": "2",
          "id": "2",
          "items": Array [
            Object {
              "product-id": "B102",
              "quantity": "5",
              "total": "24.95",
              "unit-price": "4.99",
            },
          ],
          "total": "24.95",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <orderRowContainer
          order={
            Object {
              "customer-id": "2",
              "id": "2",
              "items": Array [
                Object {
                  "product-id": "B102",
                  "quantity": "5",
                  "total": "24.95",
                  "unit-price": "4.99",
                },
              ],
              "total": "24.95",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`OrderDetail Component renders correctly 1`] = `
Array [
  <header
    canNavigateBack={true}
    childView={
      <OrderDetail
        rbSheet={
          Object {
            "current": null,
          }
        }
      />
    }
    rightBarButton={
      <HeaderBarButton
        icon=""
        onPress={[Function]}
      />
    }
    title="Order Overview"
  />,
  <View
    style={
      Object {
        "bottom": 15,
        "left": 20,
        "position": "absolute",
        "right": 20,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4169E1",
          "borderRadius": 200,
          "justifyContent": "center",
          "margin": 5,
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "alignItems": "center",
            "color": "#ffffff",
            "fontSize": 16,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": undefined,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome5FreeSolid",
              },
            ]
          }
        >
          
           ORDER
        </Text>
      </Text>
    </View>
  </View>,
]
`;
